calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,258 @@
+    builder: &mut ir::Builder,
+    asmt: ir::Assignment,
+    group: ir::WRC<ir::Group>,
+) -> ir::Assignment {
+    let dst = ir::Port {
+        name: ir::Id::new("done", None),
+        width: 1,
+        direction: ir::Direction::Input,
+        parent: ir::PortParent::Group(group),
+        attributes: ir::Attributes::default(),
+    };
+    let dst_ref = Rc::new(RefCell::new(dst));
+    builder.build_assignment(dst_ref, asmt.src, *asmt.guard)
+}
+
+//Given asmt a.go = !a.done ? 1'd1 returns a.go = 1'd1.
+fn make_go_const(
+    builder: &mut ir::Builder,
+    asmt: ir::Assignment,
+) -> ir::Assignment {
+    let con = builder.add_constant(1, 1);
+    let src = ir::Port {
+        name: ir::Id::new("const 1", None),
+        width: 1,
+        direction: ir::Direction::Output,
+        parent: ir::PortParent::Cell(ir::WRC::from(&con)),
+        attributes: ir::Attributes::default(),
+    };
+    let src_ref = Rc::new(RefCell::new(src));
+    builder.build_assignment(asmt.dst, src_ref, ir::Guard::True)
+}
+
+impl Visitor for GroupToSeq {
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        let groups: Vec<ir::RRC<ir::Group>> = comp.groups.drain().collect();
+        let mut builder = ir::Builder::new(comp, sigs);
+        for g in groups.iter() {
+            let mut group = g.borrow_mut();
+            let group_name = group.name().clone();
+
+            //builds ordering. If it cannot build a valid linear ordering of length 2,
+            //then returns None, and we stop.
+            let mut order_analysis = OrderAnalysis::default();
+            let (first, second) =
+                match order_analysis.get_ordering(&group.assignments) {
+                    None => continue,
+                    Some(order) => order,
+                };
+
+            //If not all assignments either a) write to a non-combinational cell or
+            //b) write to group[done], then stops.
+            if !group
+                .assignments
+                .iter()
+                .all(OrderAnalysis::is_orderable_assignment)
+            {
+                continue;
+            }
+
+            //Will hold the writes to the first and second cell respectively, excluding the go-done asmt.
+            let (mut fst_asmts, mut snd_asmts): (
+                Vec<ir::Assignment>,
+                Vec<ir::Assignment>,
+            ) = (Vec::new(), Vec::new());
+
+            //Holds the go-done assignment, i.e. a.go = b.done
+            let mut go_done_asmt: Option<ir::Assignment> = None;
+
+            //Holds the first "go" assignment, *if* it is in the form a.go = !a.done ? 1'd1
+            let mut first_go_asmt: Option<ir::Assignment> = None;
+
+            //Holds the group[done] = done assignment;
+            let mut group_done_asmt: Option<ir::Assignment> = None;
+
+            for asmt in group.assignments.drain(..) {
```

Comment:
Seems like `OrderAnalysis` should probably construct a struct that contains all of this information as fields

Reply:
`OrderAnalysis` only deals with `&ir::Assignment` not `ir::Assignment`. Holding these  as`&ir::Assignment` fields is giving me lifetime trouble. 

So I can just add a new method in order_assignment that takes a vec of `ir::Assignment` and calculates the fields...